### ISSUE TITLE
Building for macOS target fails when installed from website, codesigning fails with "bundle format is ambiguous"

### ISSUE THREAD

**Issue by ridiculousfish** [flutter/flutter#77173]
I am new to Flutter. I am trying to follow the steps in [Test Drive](https://flutter.dev/docs/get-started/test-drive?tab=terminal) to build a macOS app, but the `macos_assemble.sh` script is failing:

     codesign --force --verbose --sign - -- /Users/peter/flutter/myapp/build/macos/Build/Products/Debug/myapp.app/Contents/Frameworks/FlutterMacOS.framework/FlutterMacOS
     /Users/peter/flutter/myapp/build/macos/Build/Products/Debug/myapp.app/Contents/Frameworks/FlutterMacOS.framework/FlutterMacOS: replacing existing signature
     /Users/peter/flutter/myapp/build/macos/Build/Products/Debug/myapp.app/Contents/Frameworks/FlutterMacOS.framework/FlutterMacOS: bundle format is ambiguous (could be app or framework)

To reproduce:

1. `flutter create myapp`
2. `cd myapp`
3. `flutter run`
4. choose macOS

System is macOS 11.2.1, Xcode 12.4 (12D4e), Intel MBP

<details>
<summary>Logs</summary>

Verbose logs are at [this gist](https://gist.github.com/ridiculousfish/96027fb6e13f31a3a1883b2c4c50e8d1)

```
flutter doctor -v
[✓] Flutter (Channel stable, 2.0.0, on macOS 11.2.1 20D74 darwin-x64, locale en-US)
    • Flutter version 2.0.0 at /Users/peter/flutter/flutter
    • Framework revision 60bd88df91 (3 hours ago), 2021-03-03 09:13:17 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] VS Code (version 1.53.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.2.1 20D74 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 88.0.4324.192

! Doctor found issues in 2 categories.
```

</details>

**Comment by jmagman** [comment]
@ridiculousfish Thanks for the logs.  I wasn't able to reproduce on 2.0 on a newly created project.  Can you attach `/Users/peter/flutter/myapp/build/macos/Build/Products/Debug/myapp.app/Contents/Frameworks/FlutterMacOS.framework/`?

**Comment by ridiculousfish** [comment]
Uploaded to [https://ridiculousfish.com/files/FlutterMacOS.framework.zip](https://ridiculousfish.com/files/FlutterMacOS.framework.zip). N.B. 41.4 MB.

**Comment by jmagman** [comment]
I'm not sure if it's an artifact from zipping, but the symlinks are missing from the framework.  Yours looks like this:
```
$ la -la Contents/Frameworks/FlutterMacOS.framework 
-rwxr-xr-x@   1 m  primarygroup  39442240 Mar  3 14:13 FlutterMacOS
drwxr-xr-x@   1 m  primarygroup       128 Mar  3 14:13 Resources
drwxr-xr-x@   4 m  primarygroup       128 Mar  3 14:13 Versions
```
But it should look like this:
```
$ la -la Contents/Frameworks/FlutterMacOS.framework 
lrwxr-xr-x   1 m  primarygroup   29 Mar  3 15:23 FlutterMacOS -> Versions/Current/FlutterMacOS
lrwxr-xr-x   1 m  primarygroup   26 Mar  3 15:23 Resources -> Versions/Current/Resources
drwxr-xr-x   4 m  primarygroup  128 Mar  2 09:50 Versions
```

Yours copied flat:
```
rsync -av --delete --filter - .DS_Store --filter - Headers --filter - Modules /Users/peter/flutter/myapp/build/macos/Build/Products/Debug/FlutterMacOS.framework
/Users/peter/flutter/myapp/build/macos/Build/Products/Debug/myapp.app/Contents/Frameworks/
[   +7 ms] sending incremental file list
[   +4 ms] FlutterMacOS.framework/
[        ] FlutterMacOS.framework/FlutterMacOS
[ +141 ms] FlutterMacOS.framework/Resources/
[        ] FlutterMacOS.framework/Resources/Info.plist
[        ] FlutterMacOS.framework/Resources/icudtl.dat
[   +3 ms] FlutterMacOS.framework/Versions/
[        ] FlutterMacOS.framework/Versions/A/
[        ] FlutterMacOS.framework/Versions/A/FlutterMacOS
[ +136 ms] FlutterMacOS.framework/Versions/A/Resources/
[        ] FlutterMacOS.framework/Versions/A/Resources/Info.plist
[        ] FlutterMacOS.framework/Versions/A/Resources/icudtl.dat
[   +3 ms] FlutterMacOS.framework/Versions/A/_CodeSignature/
[        ] FlutterMacOS.framework/Versions/A/_CodeSignature/CodeResources
[        ] FlutterMacOS.framework/Versions/Current/
[        ] FlutterMacOS.framework/Versions/Current/FlutterMacOS
[ +148 ms] FlutterMacOS.framework/Versions/Current/Resources/
[        ] FlutterMacOS.framework/Versions/Current/Resources/Info.plist
[        ] FlutterMacOS.framework/Versions/Current/Resources/icudtl.dat
[   +3 ms] FlutterMacOS.framework/Versions/Current/_CodeSignature/
[        ] FlutterMacOS.framework/Versions/Current/_CodeSignature/CodeResources
```
but it should look like
```
[   +6 ms] building file list ... done
[   +2 ms] FlutterMacOS.framework/
[        ] FlutterMacOS.framework/FlutterMacOS -> Versions/Current/FlutterMacOS
[        ] FlutterMacOS.framework/Resources -> Versions/Current/Resources
[        ] FlutterMacOS.framework/Versions/
[        ] FlutterMacOS.framework/Versions/Current -> A
[        ] FlutterMacOS.framework/Versions/A/
[        ] FlutterMacOS.framework/Versions/A/FlutterMacOS
[ +140 ms] FlutterMacOS.framework/Versions/A/Resources/
[        ] FlutterMacOS.framework/Versions/A/Resources/Info.plist
[        ] FlutterMacOS.framework/Versions/A/Resources/icudtl.dat
[   +3 ms] FlutterMacOS.framework/Versions/A/_CodeSignature/
[        ] FlutterMacOS.framework/Versions/A/_CodeSignature/CodeResources
```

1. What's the output of `ls -la /Users/peter/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework`
2. How did you install Flutter?

**Comment by jmagman** [comment]
It looks like the download from the website is also missing https://storage.googleapis.com/flutter_infra/releases/stable/macos/flutter_macos_2.0.0-stable.zip symlinks:

```
-rwxr-xr-x@  1 m  primarygroup  39442240 Feb 25 17:11 FlutterMacOS
drwxr-xr-x@ 14 m  primarygroup       448 Feb 25 15:05 Headers
drwxr-xr-x@  3 m  primarygroup        96 Feb 25 15:05 Modules
drwxr-xr-x@  4 m  primarygroup       128 Feb 25 15:05 Resources
drwxr-xr-x@  4 m  primarygroup       128 Feb 25 15:07 Versions
```
should be:
```
lrwxr-xr-x   1 m  primarygroup   29 Mar  3 15:32 FlutterMacOS -> Versions/Current/FlutterMacOS
lrwxr-xr-x   1 m  primarygroup   24 Mar  3 15:32 Headers -> Versions/Current/Headers
lrwxr-xr-x   1 m  primarygroup   24 Mar  3 15:32 Modules -> Versions/Current/Modules
lrwxr-xr-x   1 m  primarygroup   26 Mar  3 15:32 Resources -> Versions/Current/Resources
drwxr-xr-x   4 m  primarygroup  128 Feb 25 15:07 Versions
```

**Comment by jmagman** [comment]
https://storage.googleapis.com/flutter_infra/flutter/40441def692f444660a11e20fac37af9050245ab/darwin-x64-release/FlutterMacOS.framework.zip has the correct symlinks.

**Comment by jmagman** [comment]
I was able to reproduce this with the FlutterMacOS.framework packaged with https://storage.googleapis.com/flutter_infra/releases/stable/macos/flutter_macos_2.0.0-stable.zip.

The workaround is to re-download the macOS engine artifacts, cleaning your Flutter project

$ rm -rf path/to/flutter/bin/cache/artifacts/engine/darwin-x64*
$ cd your/flutter/project
$ flutter clean
$ flutter run -d macos
```

@ridiculousfish
1. Did you install Flutter from the website https://flutter.dev/docs/get-started/install/macos ?  If not, how did you install it?
2. Can you confirm the project builds after trying the above workaround to re-download the malformed frameworks?

**Comment by ridiculousfish** [comment]
Thank you for the prompt reply.

1. Yes, I installed flutter by downloading flutter_macos_2.0.0-stable.zip from https://flutter.dev/docs/get-started/install/macos
2. `flutter precache --force` reports an error:
     ```
	 [   +1 ms] error:  cannot delete old /Users/peter/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework/Resources
	               Operation not permitted
	       error:  cannot delete old /Users/peter/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework/Versions/Current
	               Operation not permitted
	       error:  cannot delete old /Users/peter/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework/Headers
	               Operation not permitted
	       error:  cannot delete old /Users/peter/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework/Modules
	               Operation not permitted
     ```
                   
[Here is the log of my shell session](https://gist.githubusercontent.com/ridiculousfish/4b3c79780d0d62c3dd0dfa33056805f1/raw/0c322d01b351ed73bcc4def60eeb0dbc3263bc64/gistfile1.txt) where I start fresh by unzipping flutter, and [here is the flutter_01.log file](https://gist.github.com/ridiculousfish/f7b7217ef5f5b038fbb49b8c46906e35) which was output.

This error is also reported by `flutter run`.

If I delete `/Users/peter/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework/` manually, then `flutter run -d macos` succeeds and the app launches (hooray).

So in summary my workaround is:

    1. `unzip flutter_macos_2.0.0-stable.zip`
    2. `rm -Rf ~/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework`
    3. `flutter --precache --force`
    4. `flutter create myapp && cd myapp && flutter run -d macos`
    
Let me know if you prefer me to create a new issue for the "cannot delete old" error.

**Comment by jmagman** [comment]
@ridiculousfish Thanks for the detailed update, good to know `flutter precache --force` fails.  Could you file a new issue for that?  I suspect it's the same problem trying to replace directories with symlinks, but I'd like to confirm.

> 2. rm -Rf ~/flutter/flutter/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework

You should `rm -rf path/to/flutter/bin/cache/artifacts/engine/darwin-x64*` to also re-download the profile and release versions of the framework.

> 3. flutter --precache --force

If you manually `rm` the above you shouldn't need to run `flutter --precache --force`, it'll get re-downloaded when you `flutter run`.  But it doesn't hurt anything.

https://github.com/flutter/flutter/pull/77184 will be hotfixed to the stable branch shortly.

> I am new to Flutter.

Welcome!  Please file anything else you find, and apologies for the poor initial experience. 🙂

**Comment by christopherfujino** [comment]
I just published version 2.0.1 and can confirm that the new zipped archive has symlinks preserved: https://storage.googleapis.com/flutter_infra/releases/stable/macos/flutter_macos_2.0.1-stable.zip

Thanks @ridiculousfish for filing the detailed issue and @jmagman for the quick fix!

**Comment by sensuikan1973** [comment]
This error also occurs on dev/beta channel, now. (real example is https://github.com/sensuikan1973/pedax/runs/2033135291)

**Comment by jmagman** [comment]
@sensuikan1973 The fix will make it to beta very soon.  dev is never hotfixed, so the next dev will contain the fix.

If you hit this issue on any channel, the workaround is:
```
$rm -rf path/to/flutter/bin/cache/artifacts/engine/darwin-x64*
```
Then try building your macOS app again.
